### PR TITLE
Update calculation of char width to respect CJK chars correctly

### LINKED ISSUE #11232: [My bug report]Typographical confusion with dual-byte characters
font:

宋体(simsun.ttf)

```powershell
price  vdat sdm u  h  bws ieps  pd message loaion
-----  ---- --- -  -  --- ----  -- ------- ------
9.62 K    294 4 15 2500    1 175  GIGAT  CHIAO,

哇  dbda llm c噢 hdd   测试 123 pid 消息 位置
-  ---- --- -- ---   -- --- --- -- --
62 KM   294  4  15 2500   1 175 千  哇
```
```powershell
PS > $f;$e
price   : $39.62
vdat    : K
sdm     : 294
u       : 4
h       : 15
bws     : 2500
ieps    : 1
pd      : 175
message :  GIGAT
loaion  : CHIAO,
```
```
plad    : 968
哇    : 62
dbda : KM
llm  : 294
c噢   : 4
hdd  : 15
测试   : 2500
123  : 1
pid  : 175
消息   : 千
位置   : 哇
plad : 96
```
```
PS > Get-Host|select Name,Version,@{N="名称";E={"q"}},@{N="196";E={"间"}}|fl                                                                                          
Name    : ConsoleHost                                             
Version : 6.2.3                                                   
名称      : q                                                     
196     : 间
```
```
PSVersion                      6.2.3/7                              
PSEdition                      Core                               
GitCommitId                    6.2.3/7                              
OS                             Linux 4.9.112-perf #1 SMP 
PREEMPT… Platform                       Unix
```

----
Test string
```json
{
  "哇": "62",
  "dbda": "KM",
  "llm": "294",
  "c噢": 4,
  "hdd": 15,
  "测试": 2500,
  "123": 1,
  "pid": 175,
  "消息": "千",
  "位置": "哇",
  "plad": 96
}
```
```json
{
  "price": "39",
  "vdat": "K",
  "sdm": "294",
  "u": 4,
  "h": 15,
  "bws": 2500,
  "ieps": 1,
  "pd": 175,
  "mesage": " IGAT",
  "loaion": "CHIAO,",
  "plad": 968
}
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

Need to fix in two places to correctly calculate length of string in buffer cells if the string contains double wide chars.  When used interactively, the console host does the calculation, so need to make a method that was only used on Windows available for all platforms and use it to calculate length in buffer cells.  When formatting goes through `Out-String`, then it uses the default length calculation in line output so that needed to use the same method.  Since the code is in two different assemblies, there is two copies of the same code.

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/11232

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs [hunk 1]
```diff
@@ -2781,7 +2781,36 @@ internal static int LengthInBufferCells(string str, int offset, bool checkEscape
                 }
```
Comment (PoshChan) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: @SteveL-MSFT, your last commit had 4 failures in `PowerShell-CI-linux`
Breakpoint SDK Unit Tests.Managing breakpoints in a remote runspace via the SDK.Can enable breakpoints
```powershell
Exception calling "EnableBreakpoint" with "1" argument(s): "The background process reported an error with the following message: Remote transport error: ."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 106
106:                 $bp = $jobRunspace.Debugger.EnableBreakpoint($bp)                

```
Breakpoint SDK Unit Tests.Managing breakpoints in a remote runspace via the SDK.Can remove breakpoints
```powershell
Exception calling "GetBreakpoints" with "0" argument(s): "The runspace state is not valid for this operation."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 120
120:             foreach ($bp in $jobRunspace.Debugger.GetBreakpoints()) {

```
Breakpoint SDK Unit Tests.Managing breakpoints in a remote runspace via the SDK.Returns an empty collection when there are no breakpoints
```powershell
Exception calling "GetBreakpoints" with "0" argument(s): "The runspace state is not valid for this operation."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 126
126:             $jobRunspace.Debugger.GetBreakpoints() | Should -HaveCount 0

```
Breakpoint SDK Unit Tests.Handling errors while managing breakpoints in a remote runspace via the SDK.Error occurred in Context block
```powershell
Exception calling "SetCommandBreakpoint" with "1" argument(s): "The runspace state is not valid for this operation."
At /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1:153 char:13
```

**Comment by SteveL-MSFT** [comment]
@PoshChan please retry linux

**Comment by PoshChan** [comment]
@SteveL-MSFT, successfully started retry of `PowerShell-CI-Linux`

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs [hunk 2]
```diff
@@ -2781,7 +2781,36 @@ internal static int LengthInBufferCells(string str, int offset, bool checkEscape
...
+                (c <= 0x115f || /* Hangul Jamo init. consonants */
+                 c == 0x2329 || c == 0x232a ||
+                 (c >= 0x2e80 && c <= 0xa4cf &&
+                  c != 0x303f) || /* CJK ... Yi */
+                 (c >= 0xac00 && c <= 0xd7a3) || /* Hangul Syllables */
+                 (c >= 0xf900 && c <= 0xfaff) || /* CJK Compatibility Ideographs */
+                 (c >= 0xfe10 && c <= 0xfe19) || /* Vertical forms */
+                 (c >= 0xfe30 && c <= 0xfe6f) || /* CJK Compatibility Forms */
+                 (c >= 0xff00 && c <= 0xff60) || /* Fullwidth Forms */
+                 (c >= 0xffe0 && c <= 0xffe6));
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: For performance we could use a pattern like:
```
if ((uint)(c - 'a') <= (uint)('z' - 'a'))
```

**Comment by iSazonov** [comment]
@SteveL-MSFT  Is milestone 7.rc1 ok?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs (hunk 2 above)
Comment (vexx32) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: 🤔 I wonder if the new `System.Text.Rune` type has some kind of provision for this sort of thing.

**Comment by SteveL-MSFT** [comment]
@iSazonov the risk of this change is low, we can take it for rc1

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs (hunk 2 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: @vexx32 I believe the Rune struct only helps with encoding and not with rendering

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: @iSazonov unless it's proven to be faster, I think the current code is more understandable

Comment (iSazonov) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: The pattern is used in Core in some places like  https://source.dot.net/#System.Private.CoreLib/TextInfo.cs,564
It is 3 operations vs 2 operations.

**Review by iSazonov — APPROVED** [review]

**Pull request merged**